shell/platform/android/test/io/flutter/embedding/android/FlutterActivityAndFragmentDelegateTest.java

```diff
@@ -0,0 +1,501 @@
+
+    // --- Execute the behavior under test ---
+    // The FlutterEngine is setup in onAttach().
+    delegate.onAttach(RuntimeEnvironment.application);
+
+    // Emulate the host and call the method that we expect to be forwarded.
+    delegate.onActivityResult(0, 0, null);
+
+    // Verify that the call was forwarded to the engine.
+    verify(mockFlutterEngine.getActivityControlSurface(), times(1)).onActivityResult(any(Integer.class), any(Integer.class), any(Intent.class));
+  }
+
+  @Test
+  public void itForwardsOnUserLeaveHintToFlutterEngine() {
+    // Create the real object that we're testing.
+    FlutterActivityAndFragmentDelegate delegate = new FlutterActivityAndFragmentDelegate(spyHost);
+
+    // --- Execute the behavior under test ---
+    // The FlutterEngine is setup in onAttach().
+    delegate.onAttach(RuntimeEnvironment.application);
+
+    // Emulate the host and call the method that we expect to be forwarded.
+    delegate.onUserLeaveHint();
+
+    // Verify that the call was forwarded to the engine.
+    verify(mockFlutterEngine.getActivityControlSurface(), times(1)).onUserLeaveHint();
+  }
+
+  @Test
+  public void itSendsMessageOverSystemChannelWhenToldToTrimMemory() {
+    // Create the real object that we're testing.
+    FlutterActivityAndFragmentDelegate delegate = new FlutterActivityAndFragmentDelegate(spyHost);
+
+    // --- Execute the behavior under test ---
+    // The FlutterEngine is setup in onAttach().
+    delegate.onAttach(RuntimeEnvironment.application);
+
+    // Emulate the host and call the method that we expect to be forwarded.
+    delegate.onTrimMemory(TRIM_MEMORY_RUNNING_LOW);
+
+    // Verify that the call was forwarded to the engine.
+    verify(mockFlutterEngine.getSystemChannel(), times(1)).sendMemoryPressureWarning();
+  }
+
+  @Test
+  public void itSendsMessageOverSystemChannelWhenInformedOfLowMemory() {
+    // Create the real object that we're testing.
+    FlutterActivityAndFragmentDelegate delegate = new FlutterActivityAndFragmentDelegate(spyHost);
+
+    // --- Execute the behavior under test ---
+    // The FlutterEngine is setup in onAttach().
+    delegate.onAttach(RuntimeEnvironment.application);
+
+    // Emulate the host and call the method that we expect to be forwarded.
+    delegate.onLowMemory();
+
+    // Verify that the call was forwarded to the engine.
+    verify(mockFlutterEngine.getSystemChannel(), times(1)).sendMemoryPressureWarning();
+  }
+
+  /**
+   * Creates a mock {@link FlutterEngine}.
+   * <p>
+   * The heuristic for deciding what to mock in the given {@link FlutterEngine} is that we
+   * should mock the minimum number of necessary methods and associated objects. Maintaining
+   * developers should add more mock behavior as required for tests, but should avoid mocking
+   * things that are not required for the correct execution of tests.
+   */
+  @NonNull
+  private FlutterEngine mockFlutterEngine() {
```

Comment:
nit: Mocking out this many methods on it makes it seem like `FlutterEngine` should be a fake and not a mock at this point to me. If we really absolutely need to mock it and can't use the real thing. We've talked about it a little offline, but this looks like if we really need to be mocking all this stuff out this could use DI, preferably just basic manual constructor injected DI to start with. That way we could use as many real instances of things as possible and just stub/mock/fake the actual things that wouldn't work in a test. This is also probably a bigger refactor worth doing in a followup though.